File: Leetcode3507. Min pair removal to sort array I/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minimumPairRemoval(vector<int>& nums) {
        int count = 0;

        while (nums.size() > 1) {
            bool isAscending = true;
            int minSum = INT_MAX;
            int targetIndex = -1;

            for (int i = 0; i < nums.size() - 1; i++) {
                int pairSum = nums[i] + nums[i + 1];

                if (nums[i] > nums[i + 1]) {
                    isAscending = false;
                }

                if (pairSum < minSum) {
                    minSum = pairSum;
                    targetIndex = i;
                }
            }

            if (isAscending) break;

            count++;
            nums[targetIndex] = minSum;
            nums.erase(nums.begin() + targetIndex + 1);
        }

        return count;
    }
};


// runtime - 1ms
class Solution {
    public int minimumPairRemoval(final int[] nums) {
        int n = nums.length, count = 0;

        while(n > 1) {
            int minSum = Integer.MAX_VALUE, minIdx = -1;
            boolean decreasing = true;

            for(int i = 1; i < n; ++i) {
                if(nums[i - 1] + nums[i] < minSum) {
                    minSum = nums[i - 1] + nums[i];
                    minIdx = i - 1;
                }

                if(nums[i - 1] > nums[i])
                    decreasing = false;
            }

            if(decreasing)
                return count;

            nums[minIdx] = minSum;

            for(int i = minIdx + 1; i < n - 1; ++i)
                nums[i] = nums[i + 1];

            count++;
            n--;
        }

        return count;
    }
}

File: Leetcode3507. Min pair removal to sort array I/Solution.py
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        count = 0

        while len(nums) > 1:
            isAscending = True
            minSum = float("inf")
            targetIndex = -1

            for i in range(len(nums) - 1):
                pair_sum = nums[i] + nums[i + 1]

                if nums[i] > nums[i + 1]:
                    isAscending = False

                if pair_sum < minSum:
                    minSum = pair_sum
                    targetIndex = i

            if isAscending:
                break

            count += 1
            nums[targetIndex] = minSum
            nums.pop(targetIndex + 1)

        return count


# runtime - 1ms
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        c=0
        i=0
        n=len(nums)
        while i<n-1:
            if(nums[i]>nums[i+1]):
                mini=math.inf
                index=0
                for j in range(n-1):
                    if(mini>(nums[j]+nums[j+1])):
                        mini=nums[j]+nums[j+1]
                        index=j
                nums[index]=mini
                nums.pop(index+1)
                c=c+1
                n=n-1
                i=0
            else:
                i=i+1
        return c

# runtime - 3ms
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        def nondecreasing(a):
            for i in range(1, len(a)):
                if a[i] < a[i - 1]:
                    return False
            return True

        ops = 0
        while not nondecreasing(nums):
            min_sum = float("inf")
            idx = 0
            for i in range(len(nums) - 1):
                s = nums[i] + nums[i + 1]
                if s < min_sum:          # leftmost tie kept automatically
                    min_sum = s
                    idx = i
            nums[idx] = nums[idx] + nums[idx + 1]
            nums.pop(idx + 1)
            ops += 1
        return ops
